### PR TITLE
Reference pages for schema, synonym and data files design

### PR DESCRIPTION
**Pull request by DarrinEide**

### CODE REVIEW

File: academic-services/knowledge-exploration-service/reference-index-synonym.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+title: Synonyms
+description: Defines the file format and structure for synonym data in MAKES
+ms.topic: reference
+ms.date: 7/1/2020
+---
+
+# Synonym files
+
+Synonyms map equivalent terms that implicitly expand the scope of a query, without users having to know canonical terms. For example, given an entity with a string attribute "pet_type" having the canonical value of "canine" we could define the synonyms "dog" and "puppy" which would allow queries containing any of the terms "canine", "dog" or "puppy" to match the entity.
+
```
Comment (alch-msft) [inline academic-services/knowledge-exploration-service/reference-index-synonym.md:11]: add section header
```
# Synonym data file format
```

File: academic-services/knowledge-exploration-service/reference-index-data.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+string | string | `"Ti": "an overview of microsoft academic service mas and applications"`
+
+### Multiple attribute values
+
+To define more than one value for an attribute you can either specify the attribute multiple times, e.g.:
+
+``` JSON
+{ "RId": 123, "RId": 456 }
+{ "F": { "FN": "machine learning", "FId": 119857082 }, "F": { "FN": "artificial intelligence", "FId": 154945302 } }
+```
```
Comment (alch-msft) [inline academic-services/knowledge-exploration-service/reference-index-data.md:34]: These violates JSON's unique key rule. (preprocessor might crash?) Let's not encourage this notation.

File: academic-services/knowledge-exploration-service/reference-index-data.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+``` JSON
+{ "RId": [ 123, 456 ] }
+{ "F": [ { "FN": "machine learning", "FId": 119857082 }, { "FN": "artificial intelligence", "FId": 154945302 } ] }
+```
+
+## Entity log probability
+
+An important component of the KES top-N retrieval algorithm is the concept of entity probability, which informs KES of each entities relative importance. Entity probability can be understood as the likelihood of it being named the most important in a survey conducted on all members its network/graph ([see "Saliency: An Eigencentrality Measure for Heterogeneous Dynamic Network" in "A Review of Microsoft Academic Services for Science of Science Studies"](https://www.frontiersin.org/articles/10.3389/fdata.2019.00045/pdf)).
+
+KES allows this to be defined for an entity as a relative log probability using the optional "logprob" attribute, which is an additional attribute that is always available and not defined in the [schema](reference-index-schema.md). Given a probability *p* between 0 and 1, the corresponding log probability can be computed as log(*p*), where log() is the natural log function. When no value is specified for logprob, the default value 0 is used.
```
Comment (alch-msft) [inline academic-services/knowledge-exploration-service/reference-index-data.md:47]: weight

File: academic-services/knowledge-exploration-service/reference-index-data.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+### Academic paper entity
+
+The following entity JSON data is for an index that supports academic paper entities, and is part of the larger [MAKES customization tutorial](tutorial.md) and uses a subset ([defined in the schema reference](reference-index-schema.md#example)) of the full [MAKES academic entity schema](reference-makes-api-entity-schema.md).
+
+> [!IMPORTANT]
+> The following example is purposefully using tabular formatting of the JSON data to make it easy to understand. However in practice **this data would have no such formatting and be contained on a *single* line**.
+
+``` JSON
+{
+  "logprob": -19.425,
```
Comment (alch-msft) [inline academic-services/knowledge-exploration-service/reference-index-data.md:70]: weight

Comment (alch-msft) [inline academic-services/knowledge-exploration-service/reference-index-data.md:70]: long value

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 3 above)
Comment (alch-msft) [inline academic-services/knowledge-exploration-service/reference-index-data.md:47]: logprob -> weight

File: academic-services/knowledge-exploration-service/reference-index-schema.md [hunk 5]
```diff
@@ -0,0 +1,358 @@
+---
+title: Index schema
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:2]: I excluded this on purpose. It's not required, and trying to document it opens up a conversation about the inner-workings of how KES indexes data that I'm not clear we're ready for.

---
In reply to: [455218133](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455218133) [](ancestors = 455218133)

File: academic-services/knowledge-exploration-service/reference-index-schema.md [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+description: Defines the file format and structure for MAKES index schema
+ms.topic: reference
+ms.date: 7/1/2020
+---
+
+# Index schema files
+
+Index schema files define the attribute structure of entities in an index, including their [name](#attribute-name), [data type](#attribute-types), [supported operations](#attribute-operations) and [synonyms](reference-index-synonym.md).
+
+They are a critical part of the design of a [MAKES API](reference-makes-api.md):
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:12]: "They are a critical part of the designing of a MAKES API"?

---
In reply to: [455244855](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455244855) [](ancestors = 455244855)

[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:12]: Changed to "They play a critical part in the design of a MAKES API"

---
In reply to: [458259797](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r458259797) [](ancestors = 458259797,455244855)

File: academic-services/knowledge-exploration-service/reference-index-schema.md [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+
+* They serve as the blueprint for defining entity data in [data files](reference-index-data.md) used to build an index
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:14]: We need to standardize on our terminology. Let's go with "entity data files"

---
In reply to: [455246228](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455246228) [](ancestors = 455246228)

File: academic-services/knowledge-exploration-service/reference-index-schema.md [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+}
+```
+
+## Attribute name
+
+Attribute names must be composed of alphanumeric characters (a-z, A-Z,  0-9) or underscores, with the first character required to be an alpha character or underscore.
+
+Examples of valid names:
+
+* attributeName
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:41]: I disagree. I think we risk pigeonholing users into what is possible with KES if we *only* use Academic examples

---
In reply to: [455251968](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455251968) [](ancestors = 455251968)

File: academic-services/knowledge-exploration-service/reference-index-schema.md [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+* _attr
+* _012
+
+Examples of invalid names:
+
+* 0attr
+* $attributeName
+* attribute-name
+
+In addition names can also contain a [composite group](#composite-attributes) name scoping prefix in the form of "composite_group_name.", which indicates the attribute is a sub-attribute of the composite group. Both the composite group name and the sub-attribute name must follow the same naming rules defined above, i.e.:
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:52]: ?

---
In reply to: [455254455](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455254455) [](ancestors = 455254455)

[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:52]: I do this in the example immediately below

---
In reply to: [455254341](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455254341) [](ancestors = 455254341)

File: academic-services/knowledge-exploration-service/reference-index-schema.md [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+Name | Description | Supported operations | Example
+--- | --- | --- | ---
+blob | Internally compressed non-indexed data, useful for storing large amounts of data | | "It was the best of times, it was the worst of times ... It is a far, far better thing that I do, than I have ever done; it is a far, far better rest that I go to than I have ever known."
+composite | Composition of one or more non-composite attributes | | See [composite attribute](#composite-attributes) section
+date | Date value in YYYY-MM-DD format with supported range of 1400-01-01 to 9999-12-31 | equals, is_between | "2015-05-18"
+double | Double-precision floating-point value | equals, starts_with, is_between | 2.71828182846
+guid | Globally unique identifier value in XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX format | equals | "3d119f86-dcf0-4e35-9c01-b84832d851e9"
+int32 | Signed 32-bit integer | equals, starts_with, is_between | -1, 0, 2015
+int64 | Signed 64-bit integer | equals, starts_with, is_between | -1, 0, 8589934592
+string | String of up to 1024 characters | equals, starts_with | "an overview of microsoft academic service mas and applications"
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md:72]: Really what?

---
In reply to: [455257306](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455257306) [](ancestors = 455257306)

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 2 above)
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-data.md:34]: There is no requirement for JSON object names to be unique:
[https://www.json.org/json-en.html](https://www.json.org/json-en.html)

---
In reply to: [455279249](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455279249) [](ancestors = 455279249)

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 3 above)
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-data.md:47]: KES allows you to define an entities "weight" using either a log probability score in the field "logprob" or a weight score derived from the log probability (-1000*logprob) in the field "weight"

We can also describe the "weight" field if you'd like, however I'd argue it's adding unneeded complexity

---
In reply to: [455289149](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455289149) [](ancestors = 455289149)

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 4 above)
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-data.md:70]: ?

---
In reply to: [455280696](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455280696) [](ancestors = 455280696)

Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-data.md:70]: See my explanation above

---
In reply to: [455280592](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455280592) [](ancestors = 455280592)

File: academic-services/knowledge-exploration-service/reference-index-synonym.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
+---
+title: Synonyms
+description: Defines the file format and structure for synonym data in MAKES
+ms.topic: reference
+ms.date: 7/1/2020
+---
+
+# Synonym files
+
+Synonyms map equivalent terms that implicitly expand the scope of a query, without users having to know canonical terms. For example, given an entity with a string attribute "pet_type" having the canonical value of "canine" we could define the synonyms "dog" and "puppy" which would allow queries containing any of the terms "canine", "dog" or "puppy" to match the entity.
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-synonym.md:10]: ?

---
In reply to: [455270464](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455270464) [](ancestors = 455270464)

File: academic-services/knowledge-exploration-service/TOC.yml [hunk 12]
```diff
@@ -48,8 +50,19 @@
         href: reference-post-evaluate.md
       - name: POST Interpret
         href: reference-post-interpret.md
-  - name: Command line tool
-    href: reference-makes-command-line-tool.md
+  - name: Create custom MAKES index
+    expanded: false
+    items:
+    - name: Define schema
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/TOC.yml:56]: Moved to how-to

---
In reply to: [455247704](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455247704) [](ancestors = 455247704)

File: academic-services/knowledge-exploration-service/TOC.yml [hunk 13]
```diff
@@ -48,8 +50,19 @@
...
+      href: reference-index-schema.md
+    - name: Define synonyms
+      href: reference-index-synonym.md
+    - name: Define data
+      href: reference-index-data.md
+    - name: Build index
```
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/TOC.yml:62]: Sure

---
In reply to: [455240812](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r455240812) [](ancestors = 455240812)

File: academic-services/knowledge-exploration-service/reference-index-schema.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (alch-msft) [reply academic-services/knowledge-exploration-service/reference-index-schema.md]: Our default index won't build without using this capability. I think we need to illustrate what it does, at least at the concept level.

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 2 above)
Comment (alch-msft) [reply academic-services/knowledge-exploration-service/reference-index-data.md]: Preprocessor won't be able to handle this scenario. It currently depends on newtonsoft to parse JSON and it resolve duplicate keys by taking the last one. I'm not convinced this is a valuable scenario to add support at the moment. 
image

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 3 above)
Comment (alch-msft) [reply academic-services/knowledge-exploration-service/reference-index-data.md]: We leveraged "Rank" provided by MAG as the "Weight" field in our subgraph sample script. We should be consistent with what we explain and what we use. Let's either modify the script and leverage "logprob" or illustrate "Weight" here.

File: academic-services/knowledge-exploration-service/reference-index-data.md (hunk 2 above)
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-data.md]: We're not adding support, we're documenting [how KES works](https://github.com/microsoft/cognitive-research-technologies-docs/blob/master/knowledge-exploration-service/DataFormat.md#data-format)

File: academic-services/knowledge-exploration-service/reference-index-schema.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-index-schema.md]: Meaning it times out, or something else?

---
In reply to: [459005468](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/164#discussion_r459005468) [](ancestors = 459005468)

**Pull request merged**